spring-session-jdbc/src/main/java/org/springframework/session/jdbc/JdbcOperationsSessionRepository.java

```diff
@@ -735,11 +736,16 @@ public void setAttribute(String attributeName, Object attributeValue) {
 			if (attributeValue == null) {
 				this.delta.put(attributeName, DeltaValue.REMOVED);
 			}
-			else if (this.delegate.getAttribute(attributeName) != null) {
-				this.delta.put(attributeName, DeltaValue.UPDATED);
-			}
 			else {
-				this.delta.put(attributeName, DeltaValue.ADDED);
+				final Object existingAttributeValue = this.delegate.getAttribute(attributeName);
+				if (existingAttributeValue != null) {
+					if (existingAttributeValue == attributeValue && ClassUtils.isPrimitiveOrWrapper(attributeValue.getClass())) {
```

Comment:
Your implementation seems not correct at first glance. You have not handled `existingAttributeValue != attributeValue` which has to update as well.

```
					if (existingAttributeValue != attributeValue) {
						this.delta.put(attributeName, DeltaValue.UPDATED);
					}else if(!ClassUtils.isPrimitiveOrWrapper(attributeValue.getClass() || !attributeValue.getClass().equals(String.class))){
                        this.delta.put(attributeName, DeltaValue.UPDATED);
                    }
```